src/plugins/dashboard/public/services/dashboard_content_management/lib/migrate_dashboard_input.ts

```diff
@@ -0,0 +1,72 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  runEmbeddableFactoryMigrations,
+  EmbeddableFactoryNotFoundError,
+} from '@kbn/embeddable-plugin/public';
+import { ControlGroupInput } from '@kbn/controls-plugin/common';
+
+import { type DashboardEmbeddableService } from '../../embeddable/types';
+import { DashboardContainerInput, DashboardPanelState } from '../../../../common';
+
+/**
+ * Run Dashboard migrations clientside. We pre-emptively run all migrations for all content on this Dashboard so that
+ * we can ensure the `last saved state` which eventually resides in the Dashboard public state is fully migrated.
+ * This prevents the reset button from un-migrating the panels on the Dashboard. This also means that the migrations may
+ * get skipped at Embeddable create time - unless states with older versions are saved in the URL or session storage.
+ */
+export const migrateDashboardInput = (
+  dashboardInput: DashboardContainerInput,
+  embeddable: DashboardEmbeddableService
+) => {
+  let anyMigrationRun = false;
+  if (!dashboardInput) return dashboardInput;
+  if (dashboardInput.controlGroupInput) {
+    /**
+     * If any Control Group migrations are required, we will need to start storing a Control Group Input version
+     * string in Dashboard Saved Objects and then running the whole Control Group input through the embeddable
+     * factory migrations here.
+     */
+
+    // Migrate all of the Control children as well.
+    const migratedControls: ControlGroupInput['panels'] = {};
+
+    Object.entries(dashboardInput.controlGroupInput.panels).forEach(([id, panel]) => {
+      const factory = embeddable.getEmbeddableFactory(panel.type);
+      if (!factory) throw new EmbeddableFactoryNotFoundError(panel.type);
+      const { input: newInput, migrationRun: controlMigrationRun } = runEmbeddableFactoryMigrations(
+        panel.explicitInput,
+        factory
+      );
+      if (controlMigrationRun) anyMigrationRun = true;
+      panel.explicitInput = newInput as DashboardPanelState['explicitInput'];
+      migratedControls[id] = panel;
+    });
+  }
+  const migratedPanels: DashboardContainerInput['panels'] = {};
+  Object.entries(dashboardInput.panels).forEach(([id, panel]) => {
+    const factory = embeddable.getEmbeddableFactory(panel.type);
+    if (!factory) throw new EmbeddableFactoryNotFoundError(panel.type);
+    // run last saved migrations for by value panels only.
+    if (!panel.explicitInput.savedObjectId) {
+      const { input: newInput, migrationRun: panelMigrationRun } = runEmbeddableFactoryMigrations(
+        panel.explicitInput,
+        factory
+      );
+      if (panelMigrationRun) anyMigrationRun = true;
+      panel.explicitInput = newInput as DashboardPanelState['explicitInput'];
+    } else if (factory.latestVersion) {
+      // by reference panels are always considered to be of the latest version
```

Comment:
So a by reference panel that needs a migration would not trigger a "Save recommended"? If a user does edit and save the dashboard would we save the migrated by reference panels too?

Reply:
For the time being, by reference Embeddables should never need a migration via this system. Any embeddables that can be both by reference or by value will need to implement the same migration function in both the SO migrations, and the clientside Embeddable migrations.

So currently, no it won't trigger a "Save recommended", but if the user saves, we will save the `latestVersion` key into the Embeddable's input.

A by reference embeddable _may_ need migration from this system when we implement [drilldown migrations](https://github.com/elastic/kibana/issues/164500), but won't until then.